contrib/platform/src/com/sun/jna/platform/linux/LibC.java

```diff
@@ -96,6 +96,59 @@
         }
     }
 
+    @FieldOrder({ "fsBlockSize", "fsFragmentSize", "fsSizeInBlocks", "fsBlocksFree", "fsBlocksFreeUnpriv",
+            "fsTotalInodeCount", "fsFreeInodeCount", "fsFreeInodeCountUnpriv", "fsId", "_f_unused", "fsMountFlags",
+            "fsMaxFilenameLength", "_fSpare" })
+    class Statvfs extends Structure {
+        private static final int PADDING_SIZE = 8 / NativeLong.SIZE - 1;
+
+        public NativeLong fsBlockSize;
+        public NativeLong fsFragmentSize;
+        public NativeLong fsSizeInBlocks;
+        public NativeLong fsBlocksFree;
+        public NativeLong fsBlocksFreeUnpriv;
+        public NativeLong fsTotalInodeCount;
+        public NativeLong fsFreeInodeCount;
+        public NativeLong fsFreeInodeCountUnpriv;
+        public NativeLong fsId;
+        public int[] _f_unused = new int[PADDING_SIZE]; // 0 on 64-bit systems
+        public NativeLong fsMountFlags;
+        public NativeLong fsMaxFilenameLength;
+        public int[] _fSpare = new int[6];
```

Comment:
Could you please describe where the C structure is from? The linux manpage has this to say:

```C
           struct statvfs {
               unsigned long  f_bsize;    /* Filesystem block size */
               unsigned long  f_frsize;   /* Fragment size */
               fsblkcnt_t     f_blocks;   /* Size of fs in f_frsize units */
               fsblkcnt_t     f_bfree;    /* Number of free blocks */
               fsblkcnt_t     f_bavail;   /* Number of free blocks for
                                             unprivileged users */
               fsfilcnt_t     f_files;    /* Number of inodes */
               fsfilcnt_t     f_ffree;    /* Number of free inodes */
               fsfilcnt_t     f_favail;   /* Number of free inodes for
                                             unprivileged users */
               unsigned long  f_fsid;     /* Filesystem ID */
               unsigned long  f_flag;     /* Mount flags */
               unsigned long  f_namemax;  /* Maximum filename length */
           };.
       // Here  the types fsblkcnt_t and fsfilcnt_t are defined in <sys/types.h>.
       // Both used to be unsigned long
```

This would indicate, that a mapping with all elements mapped as NativeLong and no padding would be correct. I'm also wondering about the ints used as padding. Are you sure that are not byte[]?

Reply:
The manpage defines the structure "approximately as follows".  The header of [`sys/statvfs.h`](https://code.woboq.org/gcc/include/sys/statvfs.h.html#25) imports from [`bits/statvfs.h`](https://code.woboq.org/gcc/include/bits/statvfs.h.html#statvfs) where the structure is defined.  The relevant padding is a single 32-bit `int`:
```
#ifdef _STATVFSBUF_F_UNUSED
  int __f_unused;
#endif
```
That conditional define is based on having [32-bit wordsize](https://code.woboq.org/userspace/glibc/sysdeps/x86/bits/wordsize.h.html):
```
#if (__WORDSIZE == 32 \
    && (!defined __SYSCALL_WORDSIZE || __SYSCALL_WORDSIZE == 32))
  #define _STATVFSBUF_F_UNUSED
#endif
```
I've tested this both on 32-bit and 64-bit Linux VMs (Ubuntu 14 out-of-the box) and got sane values (255) for `f_namemax` on both systems.

Reply:
Related but not relevant to this PR, the [statvfs manpage on FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=statvfs&manpath=FreeBSD+8.1-RELEASE) is rather amusing.

Reply:
Thanks for the explanation. I should really now better to check the headers (I routinely pointed people to base windows bindings not on MSDN doc, but on the headers ...).

For `_f_unused` please bind it as a plain `int`. The field logic (taking it out of the field to be considered) still works. Then the padding calculation can be removed also.

For the structure names - I see why you named them how they are named. One thing you might want to consider is, that googling for the structures "header" names is easier when the original name is used. You could move the "speaking" names into javadoc.

Reply:
Will do. My thought with the array was hoping for future zero-length array compatibility. :)